### PR TITLE
feat: migrate existing Lambda layer resources

### PR DESCRIPTION
**Pull request by jhockett**
#### Checklist

- [ ] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  );
+
+  const layerVersions = Object.keys(layerVersionMap)
+    .map(version => parseInt(version, 10))
+    .sort((a, b) => b - a);
+
+  const permissions: LegacyPermission[] = layerVersionMap[`${_.first(layerVersions)}`]?.permissions;
+
+  if (permissions === undefined) {
+    context.print.warning('Unable to find layer permissions, falling back to default.');
```
Comment (attilah) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts:100]: I think we need to provide more information here for better DX: layer name for example as if the project has multiple layers defined this does not give enough information to the customer.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    layerConfiguration.permissions = [defaultLayerPermission];
+  } else {
+    layerConfiguration.permissions = [];
+    permissions.map(permission => {
+      switch (permission.type) {
+        case LegacyPermissionEnum.Private:
+          layerConfiguration.permissions.push({ type: PermissionEnum.Private });
+          break;
+        case LegacyPermissionEnum.AwsAccounts:
+          layerConfiguration.permissions.push({ type: PermissionEnum.AwsAccounts, accounts: permission.accounts });
```
Comment (attilah) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts:110]: do we have or need validation around `.accounts` and `.orgs`?

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+${chalk.red('Ensure your layer content is backed up!')}`);
+}
+
+function migrateAmplifyProjectFiles(layerName: string, latestLegacyHash: string, envName?: string) {
+  const projectRoot = pathManager.findProjectRoot();
+  removeLayerFromTeamProviderInfo(envName, layerName, projectRoot);
+  const meta = stateManager.getMeta(projectRoot);
+
+  if (_.get(meta, [categoryName, layerName, layerVersionMapKey], undefined)) {
+    _.set(meta, [categoryName, layerName, layerVersionMapKey], undefined);
```
Comment (attilah) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts:160]: set it undefined or delete the property?

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  _.set(meta, [categoryName, layerName, versionHash], latestLegacyHash);
+  stateManager.setMeta(projectRoot, meta);
+}
+
+export function removeLayerFromTeamProviderInfo(envName: string, layerName: string, projectRoot?: string) {
+  const nonCfnDataKey = 'nonCFNdata';
```
Comment (attilah) [inline packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts:168]: naming: `nonCFNData` ?

**Review by attilah — CHANGES_REQUESTED** [review]
Looks good in general added some questions about the implementation and other comments.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/storeResources.ts [hunk 5]
```diff
@@ -82,7 +83,7 @@ export const updateLayerArtifacts = async (
 
 export function removeLayerArtifacts(context: $TSContext, layerName: string) {
   if (isMultiEnvLayer(layerName)) {
-    removeLayerFromTeamProviderInfo(context, layerName);
+    removeLayerFromTeamProviderInfo(context.amplify.getEnvInfo().envName, layerName);
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/storeResources.ts:86]: No, this function is only called after a layer resource is removed with `amplify remove function`

File: packages/amplify-provider-awscloudformation/src/push-resources.ts [hunk 6]
```diff
@@ -1017,3 +1023,19 @@ export async function generateAndUploadRootStack(context: $TSContext, destinatio
...
+
+function rollbackLambdaLayers(layerResources: $TSAny[]) {
+  if (layerResources.length > 0) {
+    const projectRoot = pathManager.findProjectRoot();
+    const ccbMeta = stateManager.getCurrentMeta(projectRoot);
+    const meta = stateManager.getMeta(projectRoot);
+
+    layerResources.forEach(r => {
+      const layerMetaPath = ['function', r.resourceName, 'latestPushedVersionHash'];
+      const previousHash = _.get<string | undefined>(ccbMeta, layerMetaPath, undefined);
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-provider-awscloudformation/src/push-resources.ts:1035]: Correct, this should be a NOOP if there is no hash.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts (hunk 2 above)
Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts:110]: They are validated when added during `add/update function` flows.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts (hunk 4 above)
Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts:168]: It's correct as is. It's removing the existing key.

File: packages/amplify-category-function/src/provider-utils/awscloudformation/index.ts [hunk 7]
```diff
@@ -295,6 +303,10 @@ async function openEditor(
 }
 
 export function migrateResource(context: $TSContext, projectPath: string, service: ServiceName, resourceName: string) {
+  if (service !== ServiceName.LambdaFunction) {
+    throw new Error(`Could not get permission policies for unsupported service: ${service}`);
```
[the remark this answers is not in the document]
Comment (jhockett) [reply packages/amplify-category-function/src/provider-utils/awscloudformation/index.ts:307]: It's a sanity check, but I don't think it's reachable by any means today.

**Review by attilah — APPROVED** [review]
LGTM we can address the rest in a separate PR

**Pull request merged**